### PR TITLE
[lutron] initial submission for RA3 support

### PR DESCRIPTION
**Pull request by peterwoj**
# Description
Adding support for to LEAP bridge enabling device discovery with the RadioRA 3 processor. The RA3 processor returns nothing when requesting /device. Definition for bridge can also be targeted by make a request for the processor itself.

Introduces RA3 commands to fetch all devices and get details about the RadioRA 3 processor.

resolves openhab/openhab-addons#15540

### CODE REVIEW

**Comment by lsiepel** [comment]
Can you also update the readme.md with ra3 support?

**Comment by peterwoj** [comment]
Totally can and updated.

**Comment by bobadair** [comment]
Thanks @peterwoj! I haven't done much on the binding lately, but adding better RA3 support is a high priority. I've been meaning to eventually get one myself.
So on the RA3 you must request /device with the query string "where=IsThisDevice:false" in order to get a list of all devices? I'm not really clear on what the purpose of that is (a lot of things about LEAP seem unclear, unfortunately) but I wonder if it would be better to add a parameter to getDevices() rather than add the new getRA3Devices().  I guess it depends on whether query strings might be useful in other cases.

**Comment by peterwoj** [comment]
Not a problem, been using the binding for years now thank you for making it
happen.  The dimmers sold me on the RA3 system.

The initial approach I did overload the method and for whatever reason
didn’t feel right.  I can give it a go and have a peak again as I’ve become
more comfortable around the binding.

Yes, I’m guessing it’s a bug with Lutron and RA3.  The IsThisDevice is the
processor the binding is requesting from (I think).  Which is why the two
calls one for each device and one for props from the processor.  With the
changes I currently have I have all details propagated for the processor.

I think that i have a problem with fetching devices and clearing out the
mapping of device to zone.  I fetch the processor directly add it to the
mapping and then fetch the remaining devices which clears the mappings.

I’ll have a look later tonight about what I can do with the above and
update getDevices with a parameter.  I don’t have a caseta hub and cannot
check if my changes work with that system as well.

Peter

**Comment by peterwoj** [comment]
Updated to remove getRA3Devices() and looking a little closer I don't think that clearing out the RA3 Processor from zone <-> device mappings will cause any harm.  I've been running this for a couple of weeks now without problems.

File: bundles/org.openhab.binding.lutron/src/main/java/org/openhab/binding/lutron/internal/handler/LeapBridgeHandler.java
Comment (lsiepel) [inline bundles/org.openhab.binding.lutron/src/main/java/org/openhab/binding/lutron/internal/handler/LeapBridgeHandler.java:667]:
```suggestion
        isRadioRA3 = "Lutron RadioRA 3 Project".equals(project.productType);
```

Would also suggest to add a constant with this 'magic string' and add it to the binding constant file, for better maintainability. Not sure if it is used elsewhere, but check that too.

File: bundles/org.openhab.binding.lutron/src/main/java/org/openhab/binding/lutron/internal/protocol/leap/LeapMessageParser.java
Comment (lsiepel) [inline bundles/org.openhab.binding.lutron/src/main/java/org/openhab/binding/lutron/internal/protocol/leap/LeapMessageParser.java:302]:
```suggestion
```

File: bundles/org.openhab.binding.lutron/src/main/java/org/openhab/binding/lutron/internal/protocol/leap/LeapMessageParserCallbacks.java
Comment (lsiepel) [inline bundles/org.openhab.binding.lutron/src/main/java/org/openhab/binding/lutron/internal/protocol/leap/LeapMessageParserCallbacks.java:18]: Please revert as it introduces [checkstyle warnings](https://ci.openhab.org/job/PR-openHAB-Addons/17008/Static_20Code_20Analysis_20Report/)

File: bundles/org.openhab.binding.lutron/src/main/java/org/openhab/binding/lutron/internal/protocol/leap/LeapMessageParser.java
Comment (lsiepel) [inline bundles/org.openhab.binding.lutron/src/main/java/org/openhab/binding/lutron/internal/protocol/leap/LeapMessageParser.java:21]: Please revert as it introduces [checkstyle warnings](https://ci.openhab.org/job/PR-openHAB-Addons/17008/Static_20Code_20Analysis_20Report/)

**Review by lsiepel — CHANGES_REQUESTED** [review]
Left some comments to check.

File: bundles/org.openhab.binding.lutron/src/main/java/org/openhab/binding/lutron/internal/protocol/leap/LeapMessageParserCallbacks.java
Comment (peterwoj) [reply bundles/org.openhab.binding.lutron/src/main/java/org/openhab/binding/lutron/internal/protocol/leap/LeapMessageParserCallbacks.java]: resolved

File: bundles/org.openhab.binding.lutron/src/main/java/org/openhab/binding/lutron/internal/protocol/leap/LeapMessageParser.java
Comment (peterwoj) [reply bundles/org.openhab.binding.lutron/src/main/java/org/openhab/binding/lutron/internal/protocol/leap/LeapMessageParser.java]: resolved

**Review by lsiepel — APPROVED** [review]
LGTM

**Comment by peterwoj** [comment]
@bobadair any chance you can have a look at provide some comments?

**Comment by nefilim** [comment]
@bobadair any chance we could get this merged? 🙏

**Review by bobadair — APPROVED** [review]
Thanks. I don't have a way to test this with Caseta or RA2 Select at the moment. It doesn't look like it should cause any problems, though, so I think we can go ahead with it.

**Comment by peterwoj** [comment]
I currently have both running and RA2 and RA3 so can confirm it works on at least those.   It sure about caseta as I don't have one of those.

**Comment by peterwoj** [comment]
@lsiepel or @bobadair  says that Merging is blocked and waiting on review.  Is there someone else that needs to approve this before being accepted?

**Comment by lsiepel** [comment]
> @lsiepel or @bobadair says that Merging is blocked and waiting on review. Is there someone else that needs to approve this before being accepted?

All is set, just have to wait for a @openhab/add-ons-maintainers to check and merge. There is a queue so just give it some time.

File: bundles/org.openhab.binding.lutron/src/main/java/org/openhab/binding/lutron/internal/protocol/leap/LeapMessageParser.java
Comment (lolodomo) [inline bundles/org.openhab.binding.lutron/src/main/java/org/openhab/binding/lutron/internal/protocol/leap/LeapMessageParser.java:189]: Please confirm that the miss of a break here is really expected.

File: bundles/org.openhab.binding.lutron/src/main/java/org/openhab/binding/lutron/internal/protocol/leap/dto/ProjectTimestamp.java
Comment (lolodomo) [inline bundles/org.openhab.binding.lutron/src/main/java/org/openhab/binding/lutron/internal/protocol/leap/dto/ProjectTimestamp.java:18]: ProjectTimestamp

File: bundles/org.openhab.binding.lutron/src/main/java/org/openhab/binding/lutron/internal/protocol/leap/LeapMessageParser.java
Comment (peterwoj) [reply bundles/org.openhab.binding.lutron/src/main/java/org/openhab/binding/lutron/internal/protocol/leap/LeapMessageParser.java:189]: great catch, it needs it

File: bundles/org.openhab.binding.lutron/src/main/java/org/openhab/binding/lutron/internal/protocol/leap/dto/ProjectTimestamp.java
Comment (peterwoj) [reply bundles/org.openhab.binding.lutron/src/main/java/org/openhab/binding/lutron/internal/protocol/leap/dto/ProjectTimestamp.java]: oops, copy pasta

**Review by lolodomo — APPROVED** [review]
LGTM, thank you

**Comment by peterwoj** [comment]
Thank you, happy to have these changes in.

**Comment by bobadair** [comment]
Thanks @peterwoj!

**Comment by nefilim** [comment]
hi @peterwoj @bobadair - I'm wondering if you have noticed the following behaviour (I use Ra3 fwiw):

Given an occupancy sensor within an Area - not associated with a Zone though (I want to control motion activated devices through software).
When subscribing to said Area Status (for an occupancy sensor updates), after receiving an OneAreaStatus ReadResponse, immediately turning around and publishing a CreateRequest to the Zone commandprocessor to turn on a dimmer ... there's a significant delay, 4 - 6 seconds for the light to actually turn on and receive the CreateResponse.

However, when turning on a dimmer in isolation (not colocated with other events) it's usually around 50-100ms - pretty instantaneous. 

This situation appears to be 100% repeatable,  just wondering if you have seen the same and is something inherent in the Lutron system or if it's something unique to my situation?

**Comment by bobadair** [comment]
> hi @peterwoj @bobadair - I'm wondering if you have noticed the following behaviour (I use Ra3 fwiw):
> 
> Given an occupancy sensor within an Area - not associated with a Zone though (I want to control motion activated devices through software). When subscribing to said Area Status (for an occupancy sensor updates), after receiving an OneAreaStatus ReadResponse, immediately turning around and publishing a CreateRequest to the Zone commandprocessor to turn on a dimmer ... there's a significant delay, 4 - 6 seconds for the light to actually turn on and receive the CreateResponse.
> 
> However, when turning on a dimmer in isolation (not colocated with other events) it's usually around 50-100ms - pretty instantaneous.
> 
> This situation appears to be 100% repeatable, just wondering if you have seen the same and is something inherent in the Lutron system or if it's something unique to my situation?

Hi. I don't have a Ra3 system yes, so I can't really comment on that. Maybe @peterwoj can, or you can try asking in the OpehHAB forum. That does seem like a very long delay for any Lutron system, though. I would try playing with the timing of the CreateRequest (e.g. introducing different delays before sending it) and see how that effects the behavior.

**Comment by peterwoj** [comment]
I just got back from a long trip might be able to look at this when I get home.  Mind sending over some model numbers on the sensor?

**Comment by nefilim** [comment]
Thanks @bobadair and @peterwoj - I thought about the same to play with a delay between event & command. I'm on the road this weekend back Sunday night but here's some data from the device definition: 

```
       "DeviceType":"RPSWallMountedOccupancySensor",
         "AssociatedArea":{
            "href":"/area/32"
         },
         "AssociatedControlStation":{
            "href":"/controlstation/746"
         },
         "href":"/device/748",
         "SerialNumber":xxx,
         "Parent":{
            "href":"/project"
         },
         "ModelNumber":"LRF2-OKLB-P",
         "LinkNodes":[
            {
               "href":"/device/748/linknode/776"
            }
         ],
         "DeviceClass":{
            "HexadecimalEncoding":"6080101"
         },
         "AddressedState":"Addressed",
         "AssociatedLink":{
            "href":"/link/226"
         }
```

**Comment by peterwoj** [comment]
Cool, that's the same sensor I have.  I'll try and see if I can get to this soon.  Long flight plus border delays are telling me that I can have a look in the next couple of days.  

I did see some odd behavior with the sensor with even just using Lutron programming.  If the sensor turned the lights on and I turned them off there seemed to be a long "reset" needed before the sensor would trigger the lights back to on.

**Comment by nefilim** [comment]
I see some comments in similar vein over here: https://community.home-assistant.io/t/automation-triggers-quickly-executes-slowly-1-3-seconds-each-piece-is-fast-on-its-own-0-01-seconds/540329/12 😳 
I'll use a Hue motion sensor tomorrow and see if/how that improves end to end latency

**Comment by nefilim** [comment]
hi @peterwoj - I'm cautiously optimistic here, I've upgraded my processor from `23.06.24f000` to `23.07.15f000` and the situation appears to be massively improved, received a motion event and got confirmation of dimmed level all in < 60ms 🎉 

I have no idea if this is a result of a fix from Lutron or from restarting the processor - I'd be very curious if you see any of this behaviour and if so, could restart the processor without updating to narrow down a "resolution". 

Thanks! 
Peter

ps. would be great to have a chat sometime and hear about your experience with RadioRA3 so far in the context of home automation

**Pull request merged**
